File: Old Java Solutions/Misc/Number of Islands.java
class Solution {

    // Iterate through array, for each 1, increment count and clear entire island
    
    public int numIslands(char[][] grid) {

        int count = 0;
        
        for (int i = 0; i < grid.length; i ++) {
            for (int j = 0; j < grid[0].length; j ++) {

                if (grid[i][j] == '1') {
                    count ++;
                    clearIslands(grid, i, j);
                }

            }
        }

        return count;
    }

    public void clearIslands(char[][] grid, int currRow, int currCol) {

        if (currRow >= grid.length || currCol >= grid[0].length || currRow < 0 || currCol < 0) {
            return;
        }

        if (grid[currRow][currCol] == '1') {
            grid[currRow][currCol] = '0';
        } else {
            return;
        }

        clearIslands(grid, currRow + 1, currCol);
        clearIslands(grid, currRow, currCol + 1);
        clearIslands(grid, currRow - 1, currCol);
        clearIslands(grid, currRow, currCol - 1);

    }
}

File: Graphs/Number of Islands.py
class Solution:

    dx = [1, -1, 0, 0]
    dy = [0, 0, 1, -1]

    def numIslands(self, grid: List[List[str]]) -> int:
            
        if not grid or not grid[0]:
            return 0

        num_islands = 0

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    num_islands += 1
                    self.dfsFill(grid, i, j)

        return num_islands

    def dfsFill(self, grid, i, j):

        if not self.inBounds(i, j, grid):
            return

        if grid[i][j] == '0':
            return
        
        grid[i][j] = '0'

        for di in range(len(self.dx)):

            self.dfsFill(grid, i + self.dx[di], j + self.dy[di])

    def inBounds(self, i, j, grid):

        return i >= 0 and j >= 0 and i < len(grid) and j < len(grid[0])
